PyInstaller/depend/analysis.py

```diff
@@ -764,12 +764,38 @@ def get_bootstrap_modules():
         mod = __import__(mod_name)  # C extension.
         if hasattr(mod, '__file__'):
             loader_mods.append((mod_name, os.path.abspath(mod.__file__), 'EXTENSION'))
+
+    # modify a copy of pyiboot01_bootstrap by related arguments
+    pyiboot01_name = 'pyiboot01_bootstrap.py'
+    pyiboot01_preserved_envs = pyiboot01_preserved_envs or []
+    pyiboot01_source = os.path.join(loaderpath, pyiboot01_name)
+    pyiboot01_context = dict(
+        pyiboot01_preserved_envs=pyiboot01_preserved_envs,
+    )
+    logger.info('Building %s with %r', pyiboot01_name, pyiboot01_context)
+    with open(pyiboot01_source, 'r') as f:
+        pyiboot01_content = f.read()
+    pyiboot01_content = render_pyfile_template(pyiboot01_content, pyiboot01_context)
+    pyiboot01_built = os.path.join(workpath, pyiboot01_name)
+    with open(pyiboot01_built, 'w') as f:
+        f.write(pyiboot01_content)
+
     # NOTE:These modules should be kept simple without any complicated dependencies.
     loader_mods +=[
         ('struct', os.path.abspath(mod_struct.__file__), 'PYMODULE'),
         ('pyimod01_os_path', os.path.join(loaderpath, 'pyimod01_os_path.pyc'), 'PYMODULE'),
         ('pyimod02_archive',  os.path.join(loaderpath, 'pyimod02_archive.pyc'), 'PYMODULE'),
         ('pyimod03_importers',  os.path.join(loaderpath, 'pyimod03_importers.pyc'), 'PYMODULE'),
-        ('pyiboot01_bootstrap', os.path.join(loaderpath, 'pyiboot01_bootstrap.py'), 'PYSOURCE'),
+        ('pyiboot01_bootstrap', pyiboot01_built, 'PYSOURCE'),
     ]
     return loader_mods
+
+
+def render_pyfile_template(content, context):
+    """Render pyfile as template, template var format: '{{ var_name }}'
+    Note the single quotes are included and will be replaced together.
+    """
+    # type:(str, dict) -> str
+    for k, v in context.items():
+        content = re.sub(r"'\{\{ " + k + r" \}\}'", repr(v), content)
+    return content
```

Comment:
I suggest `re.sub(re.escape("'{{%s}}'" % k), …`, which is easier to read

Reply:
Another idea: What about  *not* modifing a copy of pyiboot01_bootstrap, but to generate a *new* module e.g. `pyimod04_variables`, which currently only defines PRESERVED_ENVS?

Reply:
> I suggest re.sub(re.escape("'{{%s}}'" % k), …, which is easier to read

Update as suggested

> What about not modifing a copy of pyiboot01_bootstrap, but to generate a new module e.g. pyimod04_variables, which currently only defines PRESERVED_ENVS?

Considering that a new file adds more overhead, and a generated new file is more or less the same thing with a modified existing file (which means we won't do anything less in the code), I think using pyiboot01_bootstrap is enough. Also manipulating environment is as dangerous and important as updating sys module, they should sit together as company :D